beacon-chain/rpc/beacon/validators.go

```diff
@@ -934,22 +934,52 @@ func (bs *Server) GetValidatorPerformance(
 ) (*ethpb.ValidatorPerformanceResponse, error) {
 	validatorSummary := state.ValidatorSummary
 
-	reqPubKeysCount := len(req.PublicKeys)
-	beforeTransitionBalances := make([]uint64, 0, reqPubKeysCount)
-	afterTransitionBalances := make([]uint64, 0, reqPubKeysCount)
-	effectiveBalances := make([]uint64, 0, reqPubKeysCount)
-	inclusionSlots := make([]uint64, 0, reqPubKeysCount)
-	inclusionDistances := make([]uint64, 0, reqPubKeysCount)
-	correctlyVotedSource := make([]bool, 0, reqPubKeysCount)
-	correctlyVotedTarget := make([]bool, 0, reqPubKeysCount)
-	correctlyVotedHead := make([]bool, 0, reqPubKeysCount)
-	missingValidators := make([][]byte, 0, reqPubKeysCount)
+	responseCap := len(req.Indices) + len(req.PublicKeys)
+	beforeTransitionBalances := make([]uint64, 0, responseCap)
+	afterTransitionBalances := make([]uint64, 0, responseCap)
+	effectiveBalances := make([]uint64, 0, responseCap)
+	inclusionSlots := make([]uint64, 0, responseCap)
+	inclusionDistances := make([]uint64, 0, responseCap)
+	correctlyVotedSource := make([]bool, 0, responseCap)
+	correctlyVotedTarget := make([]bool, 0, responseCap)
+	correctlyVotedHead := make([]bool, 0, responseCap)
+	missingValidators := make([][]byte, 0, responseCap)
 
 	headState, err := bs.HeadFetcher.HeadState(ctx)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
 	}
 
+	// Get performance based on provided validator indices.
+	// Also track missing validators using public keys.
+	for _, idx := range req.Indices {
+		val, err := headState.ValidatorAtIndex(idx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "could not get validator: %v", err)
+		}
+		if idx >= uint64(len(validatorSummary)) {
+			// Not listed in validator summary yet; treat it as missing.
+			missingValidators = append(missingValidators, val.PublicKey)
+			continue
+		}
+		currentEpoch := helpers.CurrentEpoch(headState)
+		if !helpers.IsActiveValidator(val, currentEpoch) {
+			// Inactive validator; treat it as missing.
+			missingValidators = append(missingValidators, val.PublicKey)
+			continue
+		}
+
+		summary := validatorSummary[idx]
+		effectiveBalances = append(effectiveBalances, summary.CurrentEpochEffectiveBalance)
+		beforeTransitionBalances = append(beforeTransitionBalances, summary.BeforeEpochTransitionBalance)
+		afterTransitionBalances = append(afterTransitionBalances, summary.AfterEpochTransitionBalance)
+		inclusionSlots = append(inclusionSlots, summary.InclusionSlot)
+		inclusionDistances = append(inclusionDistances, summary.InclusionDistance)
+		correctlyVotedSource = append(correctlyVotedSource, summary.IsPrevEpochAttester)
+		correctlyVotedTarget = append(correctlyVotedTarget, summary.IsPrevEpochTargetAttester)
```

Comment:
Hello! Thanks for doing this! :) 
Theres quite a bit of duplicate code here, maybe you can make an array of all the indexes prior to getting the info from the summary?

Reply:
Yes. I can do that.